File: src/Entity/Window/Tab/Address/Navbar.php
<?php

declare(strict_types=1);

namespace Yggverse\Yoda\Entity\Window\Tab\Address;

use \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\Base;
use \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\Go;
use \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\History;
use \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\Request;

class Navbar
{
    public \GtkBox $gtk;

    public \Yggverse\Yoda\Entity\Window\Tab\Address $address;

    public \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\Base $base;
    public \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\Go $go;
    public \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\History $history;
    public \Yggverse\Yoda\Entity\Window\Tab\Address\Navbar\Request $request;

    // Defaults
    private int $_margin = 8;

    public function __construct(
        \Yggverse\Yoda\Entity\Window\Tab\Address $address
    ) {
        $this->address = $address;

        // Init navbar area
        $this->gtk = new \GtkBox(
            \GtkOrientation::HORIZONTAL
        );

        $this->setMargins(
            $this->_margin
        );

        // Append base button
        $this->base = new Base(
            $this
        );

        $this->gtk->add(
            $this->base->gtk
        );

        // Append history buttons group
        $this->history = new History(
            $this
        );

        $this->gtk->add(
            $this->history->gtk
        );

        // Append request entry, fill empty space
        $this->request = new Request(
            $this
        );

        $this->gtk->pack_start(
            $this->request->gtk,
            true,
            true,
            0
        );

        // Append go button
        $this->go = new Go(
            $this
        );

        $this->gtk->add(
            $this->go->gtk
        );
    }

    public function setMargins(
        ?int $value
    ): void
    {
        $this->gtk->set_margin_top(
            $value ?? $this->_margin
        );

        $this->gtk->set_margin_bottom(
            $value ?? $this->_margin
        );

        $this->gtk->set_margin_start(
            $value ?? $this->_margin
        );

        $this->gtk->set_margin_end(
            $value ?? $this->_margin
        );

        $this->gtk->set_spacing(
            $value ?? $this->_margin
        );
    }
}

File: src/app/browser/main/tab/data/navbar.cpp
#include "navbar.hpp"
#include "navbar/base.hpp"
#include "navbar/bookmark.hpp"
#include "navbar/history.hpp"
#include "navbar/request.hpp"
#include "navbar/update.hpp"

#include "../../../../../lib/url.hpp"

using namespace app::browser::main::tab::data;

Navbar::Navbar()
{
    // Init container
    set_orientation(
        Gtk::Orientation::HORIZONTAL
    );

    set_spacing(
        SPACING
    );

    set_margin_top(
        MARGIN
    );

    set_margin_start(
        MARGIN
    );

    set_margin_end(
        MARGIN
    );

    set_margin_bottom(
        MARGIN
    );

    // Init components
    base = new navbar::Base();

        append(
            * base
        );

    history = new navbar::History();

        append(
            * history
        );

    update = new navbar::Update();

        append(
            * update
        );

    request = new navbar::Request();

        append(
            * request
        );

    bookmark = new navbar::Bookmark();

        append(
            * bookmark
        );

    // Init actions group
    action_group = Gio::SimpleActionGroup::create();

        // Define group actions
        action_group->add_action(
            "refresh",
            sigc::mem_fun(
                * this,
                & Navbar::refresh
            )
        );

    insert_action_group(
        "navbar",
        action_group
    );
}

Navbar::~Navbar() = default;

// Actions
void Navbar::refresh()
{
    // Detect request has value
    bool has_request = request->get_text_length();

    // Toggle base button sensibility
    auto url = new ::lib::Url(
        request->get_text()
    );

    base->set_sensitive(
        has_request && !url->host.empty()
                    && !url->path.empty()
    );

    delete url;

    // Toggle update button sensibility
    update->set_sensitive(
        has_request
    );
}

// Getters
Glib::ustring Navbar::get_request_value()
{
    return request->get_text();
}